sql: add support for re-using lease cache for audit logging

fqazi opened the pull request:
Previously, when a query needed to resolve table names for audit we would always take the hit of a KV call. This could be problematic since we could incur extra round trips on multi-region clusters. To address this, this patch will add support for using the lease cache for these looks up and eliminate any extra round trips.

Fixes: #99475

Release note (performance improvement): audit logging could incur extra latency when resolving table/view/sequence names.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/99502)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
Do you have a clear understanding as to why we were going to the key-value store? It's not obvious to me how this works? Don't we generally consult leases before going to read the descriptors?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99502#-:-NRJjLrwEifnB44gJIy8:b-wawngr)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99502)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schema_resolver.go` line 226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99502#-NRJjbUH8ioFX2UwPW4l:-NRJjbUH8ioFX2UwPW4m:bwhqbg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/784f75ba3fb3f8fc03151f45dd0a15fbb2a4d4e1/pkg/sql/schema_resolver.go#L226)):*
> ```Go
> 	ctx context.Context, desc catalog.TableDescriptor,
> ) (*tree.TableName, error) {
> 	// When getting the fully qualified name, attempt to use the lease cache
> ```

This says we're going to go to the KV store, but it doesn't say *why* we need to use this flag to not get a lease. This makes the comment confusing. By default we do use leases. 


<!-- Sent from Reviewable.io -->

fqazi commented:
> Do you have a clear understanding as to why we were going to the key-value store? It's not obvious to me how this works? Don't we generally consult leases before going to read the descriptors?

We only consult them if withoutLeased is not set and this code path always intentionally was setup to avoid leased descriptors. I think this because name resolution is also done for alters

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99502)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schema_resolver.go` line 226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99502#-NRJjbUH8ioFX2UwPW4l:-NRJll6d1z8sV4JBIAat:b-axjwym) ([raw file](https://github.com/cockroachdb/cockroach/blob/784f75ba3fb3f8fc03151f45dd0a15fbb2a4d4e1/pkg/sql/schema_resolver.go#L226)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This says we're going to go to the KV store, but it doesn't say *why* we need to use this flag to not get a lease. This makes the comment confusing. By default we do use leases. 
</blockquote></details>

ok, I see now that `ByID` is `ByIDWithoutLeased` and what I think we want is just `ByIDWithLeased`. Maybe we should rename `ByID` to `ByIDWithoutLeased` to make it more explicit. It seems scary to me to add a new lookup mode.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99502)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[`pkg/sql/schema_resolver.go` line 226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99502#-NRJjbUH8ioFX2UwPW4l:-NRJlbLl5lty4noBoyA7:b-jvohzx) ([raw file](https://github.com/cockroachdb/cockroach/blob/784f75ba3fb3f8fc03151f45dd0a15fbb2a4d4e1/pkg/sql/schema_resolver.go#L226)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

ok, I see now that `ByID` is `ByIDWithoutLeased` and what I think we want is just `ByIDWithLeased`. Maybe we should rename `ByID` to `ByIDWithoutLeased` to make it more explicit. It seems scary to me to add a new lookup mode.
</blockquote></details>

I'm gonna leave the second part with the rename as a separate PR + issue (I'll post it here).  That also thew me off and should be explicit.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99502)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[`pkg/sql/schema_resolver.go` line 226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99502#-NRJjbUH8ioFX2UwPW4l:-NRJpR52DV-ifEynaHeM:b-31792h) ([raw file](https://github.com/cockroachdb/cockroach/blob/784f75ba3fb3f8fc03151f45dd0a15fbb2a4d4e1/pkg/sql/schema_resolver.go#L226)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I'm gonna leave the second part with the rename as a separate PR + issue (I'll post it here).  That also thew me off and should be explicit.
</blockquote></details>

Opened issue #99506, which I'll use to just rename this interface


<!-- Sent from Reviewable.io -->

fqazi committed:
sql: add support for re-using lease cache for audit logging

Previously, when a query needed to resolve table names for
audit we would always take the hit of reading the descriptor KV.
This was problematic and incur round trips, which are expensive
on multi-region clusters. To address this, this patch
will add support for using the lease cache for these
looks up and eliminate any extra round trips.

Fixes: #99475

Release note (performance improvement): audit logging could
incur extra latency when resolving table/view/sequence names.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod cleaning up the testing. 

> Reviewed 2 of 6 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99502#-:-NRJqzTo17MN3Pa5gvza:btb2ap4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99502#-NRJr8I62llArzkBeymj:-NRJr8I62llArzkBeymk:b-dj28aq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e5cc8176a5145c6e5d89f29e06683c48326001a/pkg/bench/rttanalysis/testdata/benchmark_expectations#L2)):*
> ```
> exp,benchmark
> 0," Audit/select_from_an_audit_table"
> ```

something seems off here -- why does the good one have a space?
___
*[`pkg/bench/rttanalysis/audit_bench_test.go` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99502#-NRJrBnq8gGmV4it0dPH:-NRJrBnq8gGmV4it0dPI:b-j79068) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e5cc8176a5145c6e5d89f29e06683c48326001a/pkg/bench/rttanalysis/audit_bench_test.go#L21)):*
> ```Go
> 			Name: "select from an audit table",
> 			Setup: `CREATE TABLE audit_table(a INT);
> 							ALTER TABLE audit_table EXPERIMENTAL_AUDIT SET READ WRITE;`,
> ```

Perhaps the test should lease the descriptor in setup by running the select statement also in the setup -- then we should see something like one round-trip, no?


<!-- Sent from Reviewable.io -->

fqazi commented:
@ajwerner TFTR!

bors r+
